ejb3/src/main/java/org/jboss/as/ejb3/timerservice/TimerServiceImpl.java

```diff
@@ -594,10 +594,48 @@ public void persistTimer(final TimerImpl timer, boolean newTimer) {
                     EJB3_TIMER_LOGGER.timerPersistenceNotEnable();
                     return;
                 }
+
+                final ContextTransactionManager transactionManager = ContextTransactionManager.getInstance();
+                Transaction clientTX = transactionManager.getTransaction();
                 if (newTimer) {
-                    timerPersistence.getValue().addTimer(timer);
+                    if( clientTX == null ){
+                        transactionManager.begin();
+                    }
+                    try {
+                        timerPersistence.getValue().addTimer(timer);
+                        if(clientTX == null) transactionManager.commit();
+                    } catch (Exception e){
+                        if(clientTX == null) {
+                            try {
+                                transactionManager.rollback();
+                            } catch (Exception ee){
+                                EjbLogger.EJB3_TIMER_LOGGER.timerUpdateFailedAndRollbackNotPossible(ee);
+                            }
+                        }
+                        throw e;
+                    }
                 } else {
-                    timerPersistence.getValue().persistTimer(timer);
+                    if( clientTX != null ) transactionManager.suspend();
```

Comment:
Timer updates should be transactional, so if the outer transaction fails the timer should not be updated.

Reply:
Only for the new Timers. Otherwise - it's an internal timer processing transaction, which MUST be finished independently of the timer's own activity.

Reply:
So I think there are two cases where this is called:
1) Cancellation, which should definitely be transactional (as if the TX rolls back the timer should not be cancelled)
2) Updates after the timer has run, which will not have an outer transaction.

For 1) I don't think it is possible to retry, for 2) retry can be attempted however there should not be an outer transaction to suspend, so I don't think you want to suspend the tx in either case

Reply:
You are right!

For the cancellation, I'm just trying to avoid a special case, related to the way, how the Oracle DB process synchronized transaction isolation. It just throws an SQL Exception on any attempt to make concurrent changes. In that case, client code is not able to cancel the timer, if it not lucky enough to stop it while some other cluster node is about to update it. BTW, I am not sure, but in case of Non-persistent Timers - timer cancellation is not transactional and it will be canceled, even if the outer transaction will be failed. So, why make any difference in here?

Reply:
I've removed TX management for the client's invocations in last commits.